### PR TITLE
Nintey degree turns

### PR DESCRIPTION
**Pull request by Wirt4**
This pull request introduces significant changes to the raycasting game engine, focusing on refactoring the `Player` class, improving the `GameMap` structure, and enhancing the rendering and input handling logic. The updates aim to increase modularity, simplify player movement and rotation, and improve test coverage for key components.

### Refactoring and Modularization

* **Player Class Implementation**: The `Player` class is now implemented as an independent module with methods for movement and rotation. It includes a new `speed` property and integrates angle normalization for consistent rotation behavior.
* **Player Interface**: A new `PlayerInterface` has been added, defining the contract for player-related properties and methods (`rotate`, `move`, `x`, `y`, and `angle`).

### GameMap Enhancements

* **GameMap Constructor Refactor**: The `GameMap` constructor now accepts a `PlayerInterface` instance and `Dimensions` object, improving flexibility and enforcing player boundary checks during initialization. [[1]](diffhunk://#diff-d0071060a183a11f4d71feab5fba42e1904595b5c8ff0ab9a4a8ef55844c1709L15-R34) [[2]](diffhunk://#diff-d0071060a183a11f4d71feab5fba42e1904595b5c8ff0ab9a4a8ef55844c1709R44-R62)
* **Player Integration in GameMap**: Player-related properties (`playerPosition`, `playerAngle`) and methods (`movePlayer`, `turnPlayer`) are now delegated to the `PlayerInterface`, reducing direct manipulation of player state. [[1]](diffhunk://#diff-d0071060a183a11f4d71feab5fba42e1904595b5c8ff0ab9a4a8ef55844c1709R44-R62) [[2]](diffhunk://#diff-d0071060a183a11f4d71feab5fba42e1904595b5c8ff0ab9a4a8ef55844c1709L67-R99)

### Rendering Improvements

* **Coordinate System Adjustment**: The rendering logic has been updated to flip the vertical axis by translating and scaling the canvas context. This ensures alignment with the game's coordinate system.

### Input Handling Updates

* **Keyboard Input for Player Movement**: The mouse-based turning mechanism has been replaced with keyboard controls (`ArrowLeft` and `ArrowRight`) for player rotation, simplifying input handling.

### Test Coverage Expansion

* **GameMap and Player Tests**: Comprehensive test cases have been added for the `Player` class (`move` and `rotate` methods) and `GameMap` initialization, ensuring correctness in edge cases like player position touching walls. [[1]](diffhunk://#diff-d7d45d0c4d9e964dd8a153934039b83b051cab229f07e437e4b390394c7896d0R1-R38) [[2]](diffhunk://#diff-7735bb36592de1aa0b881f9d05d485104d221b62189ab9f8eaad9a51e6eca82aL111-R166)

These changes collectively enhance the modularity, maintainability, and usability of the raycasting engine, laying a solid foundation for future development.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -30,7 +30,8 @@
 !/tmp/storage/.keep
 
 /public/assets
-
+/public/workers
+# assumes public workers will compile according to the build process
```
Comment (Wirt4) [inline .gitignore:34]: remove comment

File: app/raycaster/src/game/game.ts [hunk 2]
```diff
@@ -18,12 +18,12 @@ class Game {
 	}
 
 	draw(renderer: RendererInterface, raycaster: RaycasterInterface, brightness: BrightnessInterface): void {
-		renderer.fillColor(ColorName.BLACK, 0.01);
+		//TODO: check if these are neessary
```
Comment (Wirt4) [inline app/raycaster/src/game/game.ts:21]: yes they are, remove comments

File: app/raycaster/src/game/game.ts [hunk 3]
```diff
@@ -18,12 +18,12 @@ class Game {
...
+		renderer.fillColor(ColorName.GREEN, 0.01); //Sanity check to avoid black screen
 		renderer.rect({ x: 0, y: 0 }, Settings.CANVAS_WIDTH, Settings.CANVAS_HEIGHT);
 
 		const rays = raycaster.getViewRays(this.map.playerAngle);
 		const wallBatches: Record<string, BatchedRect[]> = {}
-
 		// Calculate the focal length based on the field of vision
```
Comment (Wirt4) [inline app/raycaster/src/game/game.ts:27]: turn comments into function declarations to show developer intent

File: app/raycaster/src/settings.ts [hunk 4]
```diff
@@ -5,6 +5,7 @@ export enum Settings {
 	MAX_BRIGHTNESS = 60,
 	HUD_ON = 1,
 	WALL_HEIGHT = 7,
+	PLAYER_SPEED = 0,
```
Comment (Wirt4) [inline app/raycaster/src/settings.ts:8]: I know the turn is "poppy" but try a test with a speed value

File: app/raycaster/src/worker/worker.ts [hunk 5]
```diff
@@ -41,11 +39,13 @@ onmessage = (e) => {
...
 		startLoop();
 	}
-	if (msg.type === "mouseTurn") {
-		game.map.turnPlayer(msg.angleDelta);
-	}
-	if (msg.type === "KeyDown") {
-		game.map.turnPlayer
+	if (msg.type === "turn") {
+		if (msg.direction === "left") {
+			game.map.turnPlayer(Math.PI / 2) //reverse the direction because the y axis is inverted in the canvas
```
Comment (Wirt4) [inline app/raycaster/src/worker/worker.ts:44]: move both explanations to one comment

File: app/raycaster/src/game/batches.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+		this.gridBatch.push({ x: origin.x, y: origin.y, width: 1, height: 1 });
+	}
+
+	unpackWallKey(key: string): { color: ColorName, brightness: number } {
+		const [color, brightness] = key.split("_");
+		return { color: color as ColorName, brightness: Number(brightness) / 100 };
+	}
+
+	addMapWall(wall: { color: ColorName, line: { start: Coordinates, end: Coordinates } }): void {
+		const key = `${wall.color}_0.5`; // Assuming a fixed weight for simplicity
```
Comment (Wirt4) [inline app/raycaster/src/game/batches.ts:25]: remove unnecessary comment

File: app/raycaster/src/game/game.ts [hunk 7]
```diff
@@ -13,85 +14,117 @@ class Game {
...
-		const wallBatches: Record<string, BatchedRect[]> = {}
-
-		// Calculate the focal length based on the field of vision
-		const gridBatch: Array<BatchedRect> = []
-		rays.forEach((angle, i) => {
-			const { distance, color, gridHits } = this.map.castRay(angle, Settings.MAX_DISTANCE);
-			const correctedDistance = raycaster.removeFishEye(distance, angle, this.map.playerAngle);
+		const batches = this.batchRenderData(rays, raycaster, brightness);
+		this.renderWalls(batches, renderer)
+		// Draw the floor grid
```
Comment (Wirt4) [inline app/raycaster/src/game/game.ts:39]: nix the comment here

File: app/raycaster/src/game/batches.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+		const key = `${color}_${Math.round(brightness * 100)}`;
+		if (!this.wallBatches[key]) this.wallBatches[key] = [];
+		this.wallBatches[key].push({ x: origin.x, y: origin.y, width: 1, height });
+	}
+	addGridPoint(origin: Coordinates): void {
+		this.gridBatch.push({ x: origin.x, y: origin.y, width: 1, height: 1 });
+	}
+
+	unpackWallKey(key: string): { color: ColorName, brightness: number } {
+		const [color, brightness] = key.split("_");
```
Comment (Wirt4) [inline app/raycaster/src/game/batches.ts:20]: remove duplication
